libretro-common/rest/rest.c

```diff
@@ -0,0 +1,202 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <rest/rest.h>
+#include <rthreads/rthreads.h>
+
+struct rest_request_t
+{
+   struct http_request_t *http_request;
+   struct http_connection_t *http_conn;
+   struct http_t *http;
+   rest_retry_policy_t *retry_policy;
+   int retry_count;
+   time_t delay_end;
+};
+
+struct rest_retry_policy_t
+{
+   int *status_codes;
+   size_t status_codes_len;
+   time_t *delays;
+   size_t delays_len;
+};
+
+rest_retry_policy_t *rest_retry_policy_new(
+   int *status_codes,
+   size_t status_codes_len,
+   time_t *delays,
+   size_t delays_len)
+{
+   rest_retry_policy_t *retry_policy;
+
+   retry_policy = (rest_retry_policy_t *)calloc(1, sizeof(rest_retry_policy_t));
+   retry_policy->status_codes = status_codes;
+   retry_policy->status_codes_len = status_codes_len;
+   retry_policy->delays = delays;
+   retry_policy->delays_len = delays_len;
+
+   return retry_policy;
+}
+
+void rest_request_free(rest_request_t *request)
+{
+   if (!request)
+   {
+      return;
+   }
+
+   if (request->http)
+      net_http_delete(request->http);
+
+   if (request->http_conn)
+      net_http_connection_free(request->http_conn, false);
+   else if (request->http_request)
+      net_http_request_free(request->http_request);
+
+   if (request->http_request)
+   {
+      net_http_request_free(request->http_request);
+   }
+
+   free(request);
+}
+
+rest_request_t *rest_request_new(struct http_request_t *http_request)
+{
+   struct rest_request_t *request;
+
+   request = (rest_request_t *)calloc(1, sizeof(rest_request_t));
+   request->http_request = http_request;
+
+   return request;
+}
+
+void rest_request_set_header(rest_request_t *request, const char *name, const char *value, bool replace)
+{
+   net_http_request_set_header(request->http_request, name, value, replace);
+}
+
+void rest_request_set_retry_policy(
+   rest_request_t *request,
+   rest_retry_policy_t *retry_policy)
+{
+   if (request)
+   {
+      request->retry_policy = retry_policy;
+   }
+}
+
+static void _start_request(rest_request_t *request)
+{
+   request->http_conn = net_http_connection_new(request->http_request);
+   if (!request->http_conn || !net_http_connection_iterate(request->http_conn) || !net_http_connection_done(request->http_conn))
+   {
+      net_http_connection_free(request->http_conn, true);
+      request->http_conn = NULL;
+      return;
+   }
+
+   request->http = net_http_new(request->http_conn);
+   if (!request->http)
+   {
+      net_http_connection_free(request->http_conn, true);
+      net_http_delete(request->http);
+      request->http_conn = NULL;
+      request->http = NULL;
+   }
+}
+
+struct http_response_t *rest_request_execute(rest_request_t *request)
+{
+   time_t timeout;
+   size_t progress;
+   size_t total;
+
+   timeout = time(NULL) + 30;
```

Comment:
Probably want timeout threshold configurable

Reply:
This now defaults to 30s, but there is a new function so that it can be changed.